src/crystal/compiler_rt/shift.cr

```diff
@@ -0,0 +1,53 @@
+# :nodoc:
+# Ported from https://github.com/llvm/llvm-project/blob/ce59ccd04023cab3a837da14079ca2dcbfebb70c/compiler-rt/lib/builtins/ashlti3.c
+#
+# Returns: a << b
+# Precondition:  0 <= b < bits_in_tword
+fun __ashlti3(a : Int128, b : Int32) : Int128
+  low, high = a.unsafe_as(Tuple(UInt64, Int64))
```

Comment:
No other methods in compiler-rt use `unsafe_as(Tuple(_, _))` to split variables into a high/low part.

I think it would be better to stay consistent:

https://github.com/crystal-lang/crystal/blob/100391991856c53d31e3f712ff8ab5197d687146/src/crystal/compiler_rt/multi3.cr#L24-L25

https://github.com/crystal-lang/crystal/blob/100391991856c53d31e3f712ff8ab5197d687146/src/crystal/compiler_rt/divmod128.cr#L172-L173

(Also not that important but I think using unsafe_as on a big endian system could work differently, making this a bit harder to understand)

Reply:
This would be using shift in the shift implementation :p

Reply:
Yeah, but 64-bit shifting 😄

unsafe_as will probably generate the same assembly anyways

Reply:
What assembly does unsafe_as generate if it isn't using a right-shift?

Reply:
In WASM a Int128 is already represented as two separated `i64` variables. So this generates no code when compiling in release more.

```cr
fun test_func(a : UInt128) : UInt64
  low, high = a.unsafe_as(Tuple(UInt64, UInt64))
  low &+ high
end
```

Compiles to:

```wast
  (type $t25 (func (param i64 i64) (result i64)))
  (func $test_func (type $t25) (param $p0 i64) (param $p1 i64) (result i64)
    local.get $p1
    local.get $p0
    i64.add)
```